### PR TITLE
feat: add `to_json` method to google.oauth2.credentials.Credentials

### PR DESCRIPTION
**Pull request by patkasper**
This adds a `Storage` base interface class and a `file` module based on `Storage` to save and load credentials.
The `credentials` module was modified to rely on the newer `Storage` class while retaining compatibility.

### CODE REVIEW

**Comment by busunkim96** [comment]
Hi @patkasper - apologies for it taking a long time to comment on this.

I brought up credentials storage in a weekly meeting and it looks like folks are generally *opposed* to the idea of this library handling credentials storage. It may be convenient to store credentials in memory or in a file, but it isn't great from a security perspective.

There was interest in adding a method like `to_json` to make it easier to convert a credentials object back into JSON. Could you perhaps rework this PR to add just that method?

**Comment by patkasper** [comment]
I've removed the storage interface and I've kept the `to_json` method `in google.oauth2`'s `Credentials`. Please let me know what you think.

I'm happy to add the storage interface + automated loading/storing to google-auth-oauthlib, should I go ahead with that?

Thanks.

File: google/oauth2/credentials.py [hunk 1]
```diff
@@ -195,14 +223,50 @@ def from_authorized_user_file(cls, filename, scopes=None):
...
+        # as I do not see the point in it.
+
+        prep = {
+                'token':            self.token,
+                'refresh_token':    self.refresh_token,
+                'token_uri':        self.token_uri,
+                'client_id':        self.client_id,
+                'client_secret':    self.client_secret,
+                'scopes':           self.scopes,
+                'expiry':           self.expiry.timestamp() if self.expiry
```
Comment (busunkim96) [inline google/oauth2/credentials.py]: Is this the correct timestamp format?

File: google/auth/_helpers.py [hunk 2]
```diff
@@ -232,3 +241,34 @@ def unpadded_urlsafe_b64encode(value):
         Union[str|bytes]: The encoded value
     """
     return base64.urlsafe_b64encode(value).rstrip(b'=')
+
+
+def validate_file(filename):
```
[the remark this answers is not in the document]
Comment (patkasper) [reply google/auth/_helpers.py]: Indeed, that was a remnant from my earlier additions. Removing.

File: docs/reference/google.oauth2.rst [hunk 3]
```diff
@@ -14,4 +14,5 @@ Submodules
    google.oauth2.credentials
    google.oauth2.id_token
    google.oauth2.service_account
+   google.oauth2.file
```
Comment (patkasper) [inline docs/reference/google.oauth2.rst]: Remnant from earlier changes, removing.

File: docs/reference/google.oauth2.file.rst [hunk 4]
```diff
@@ -0,0 +1,7 @@
+google.oauth2.file module
+=========================
+
+.. automodule:: google.oauth2.file
+    :members:
+    :inherited-members:
```
Comment (patkasper) [inline docs/reference/google.oauth2.file.rst]: Remnant from earlier change, removing all this.

File: google/oauth2/credentials.py [hunk 5]
```diff
@@ -74,6 +75,8 @@ def __init__(self, token, refresh_token=None, id_token=None,
                 token if refresh information is provided (e.g. The refresh
                 token scopes are a superset of this or contain a wild card
                 scope like 'https://www.googleapis.com/auth/any-api').
+            expiry (Optional(float|int)): Timestamp of the expiry of the access
```
[the remark this answers is not in the document]
Comment (patkasper) [reply google/oauth2/credentials.py]: For the time being I'm deciding to simply stop exporting/importing timestamps. I'm not finding much documentation on how to store such expiry times, and defaulting to using the refresh token to obtain an access token seems to be good practice.

File: google/oauth2/credentials.py [hunk 6]
```diff
@@ -160,13 +165,12 @@ def from_authorized_user_info(cls, info, scopes=None):
 
         Args:
             info (Mapping[str, str]): The authorized user info in Google
-                format.
+                                      format.
```
[the remark this answers is not in the document]
Comment (patkasper) [reply google/oauth2/credentials.py]: Funny, there are a few occurrences where the indent is not what I had locally. Not sure what happened.

File: google/oauth2/credentials.py [hunk 7]
```diff
@@ -179,13 +183,37 @@ def from_authorized_user_info(cls, info, scopes=None):
...
+        # instead of relying solely on scopes passed as method arguments.
+        if scopes and info.get('scopes'):
+            _scopes = set(scopes+info.get('scopes'))
+        elif info.get('scopes') and not scopes:
+            _scopes = info.get('scopes')
+        elif scopes and not info.get('scopes'):
+            _scopes = scopes
+        else:
+            _scopes = None
+
```
[the remark this answers is not in the document]
Comment (patkasper) [reply google/oauth2/credentials.py]: I see. I've removed my suggestion.

File: google/oauth2/credentials.py [hunk 8]
```diff
@@ -195,14 +223,50 @@ def from_authorized_user_file(cls, filename, scopes=None):
...
-
+            google.oauth2.credentials.Credentials: The constructed credentials.
         Raises:
             ValueError: If the file is not in the expected format.
         """
         with io.open(filename, 'r', encoding='utf-8') as json_file:
             data = json.load(json_file)
             return cls.from_authorized_user_info(data, scopes)
+
+    def to_json(self, strip=None, indent=None):
```
[the remark this answers is not in the document]
Comment (patkasper) [reply google/oauth2/credentials.py]: Ok, no problem.

File: google/oauth2/credentials.py (hunk 1 above)
Comment (patkasper) [reply google/oauth2/credentials.py]: As per previous comment, dropping this for now (or maybe forever!).

File: google/oauth2/credentials.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (busunkim96) [reply google/oauth2/credentials.py]: Would you mind reverting these whitespace changes?

File: google/oauth2/credentials.py [hunk 9]
```diff
@@ -195,14 +194,41 @@ def from_authorized_user_file(cls, filename, scopes=None):
...
             data = json.load(json_file)
             return cls.from_authorized_user_info(data, scopes)
+
+    def to_json(self, strip=None):
+        """Utility function that creates JSON repr. of a Credentials object.
+
+        Args:
+            strip (Sequence[str]): Optional list of members to exclude from the
+                                   generated JSON.
+            indent (int): This parameter is passed to `json.dumps` to create
```
Comment (busunkim96) [inline google/oauth2/credentials.py]: Please remove the documentation for this arg

File: google/oauth2/credentials.py [hunk 10]
```diff
@@ -195,14 +194,41 @@ def from_authorized_user_file(cls, filename, scopes=None):
...
+            google.oauth2.credentials.Credentials: The constructed credentials.
         Raises:
             ValueError: If the file is not in the expected format.
         """
         with io.open(filename, 'r', encoding='utf-8') as json_file:
             data = json.load(json_file)
             return cls.from_authorized_user_info(data, scopes)
+
+    def to_json(self, strip=None):
+        """Utility function that creates JSON repr. of a Credentials object.
```
Comment (busunkim96) [inline google/oauth2/credentials.py]:
```suggestion
        """Utility function that creates JSON representation of a Credentials object.
```

**Comment by busunkim96** [comment]
:wave: Hi @patkasper. Would you mind resolving the merge conflicts here?

**Comment by busunkim96** [comment]
@patkasper Would you be able to run the code formatter?

```
pip install nox
nox -s blacken
```

If nox gives you trouble for any reason you can also do:
```
pip install black
black google tests
```


@michaelawyu If you're around next week could you help see this PR over the finish line? Since Patrick is an external contributor the CI does not trigger automatically. You'll have to add the label `kokoro:force-run`. Thank you!

**Comment by nicain** [comment]
@patkasper @busunkim96 I noticed this evening that I cannot round-trip a credential using "to_json" to serialize the object, and then reload.  

I hit this when deserializing my credential and attempting to use it with a Dialogflow client; in that case, I received:

```
google.api_core.exceptions.PermissionDenied: 403 Your application has authenticated using end user credentials from the Google Cloud SDK or Google Cloud Shell which are not supported by the dialogflow.googleapis.com. We recommend configuring the billing/quota_project setting in gcloud or using a service account through the auth/impersonate_service_account setting. For more information about service accounts and how to use them in your application, see https://cloud.google.com/docs/authentication/. If you are getting this error with curl or similar tools, you may need to specify 'X-Goog-User-Project' HTTP header for quota and billing purposes. For more information regarding 'X-Goog-User-Project' header, please check https://cloud.google.com/apis/docs/system-parameters.
```

The hint here was the "billing quota"; when I compared the __dict__ representations of the original and serialized-then-serialized clone of the original object, I saw that the "_quota_project_id" attribute was missing on the clone.  Sure enough, after I added this in manually, my credential works.

Was the "quota_project_id" attribute intentionally excluded from the contents of the json blob, on serialization?  If not, I am happy to work up a PR to include that attribute in the serialization (it is a kwarg of the initializaer of the Credentials class, so very easy to round-trip).

-----

One other minor note; the serialized-then-serialized clone of the original credential cannot itself be serialized with to_json if the expiry kwarg is naively set to the expiry string from the serialization dict. This happened to me when I tried to naively deserialize by just "**" unpackinging the dictionary contents into the kwargs of the initializer.  Then [this line](https://github.com/googleapis/google-auth-library-python/blob/b79b55407b31933c9a8fe6de01478fa00a33fa2b/google/oauth2/credentials.py#L414) fails because self.expiry is a string, not a datetime.  This should be pretty easy to fix with a isinstance or try-catch, and I am happy to provide a PR here as well.

What are your thoughts on these issues?

**Pull request merged**